File: Dynamic_Programming_Old/canSum-howSum-bestSum/canSum.cpp
#include <bits/stdc++.h>
using namespace std;
using ull = unsigned long long int;

map<int, ull> memo = {};

void pMemo()
{
    for (auto e : memo)
        cout << e.first << "\t" << e.second << endl;
}

// Brute Force
bool canSum(int targetSum, vector<int> numbers)
{
    if (targetSum == 0)
        return true;
    if (targetSum < 0)
        return false;

    for (auto num : numbers)
    {
        int remain = targetSum - num;
        if (canSum(remain, numbers) == true)
            return true;
    }

    return false;
}

int main()
{
    // system("cls");
    cout << canSum(7, {2, 3}) << endl;
    cout << canSum(7, {7, 7}) << endl;
    cout << canSum(7, {5, 3}) << endl;
}

File: Dynamic_Programming_Old/canSum.js
console.clear();

// CanSum Problem

// Brute Force

// const canSum = (targetSum, numbers) => {
//     if (targetSum === 0) return true;
//     if (targetSum < 0) return false;

//     for(let num of numbers){
//         const remainder = targetSum - num;
//         if (canSum(remainder, numbers) === true){
//             return true;
//         }
//     }

//     return false;
// };


// Memoization

// const canSum = (targetSum, numbers, memo = []) => {
//     if (targetSum in memo) return memo[targetSum];
//     if (targetSum === 0) return true;
//     if (targetSum < 0) return false;

//     for(let num of numbers){
//         const remainder = targetSum - num;
//         if (canSum(remainder, numbers, memo) === true){
//             memo[targetSum] = true;
//             return true;
//         }
//     }

//     memo[targetSum] = false;
//     return false;
// };


// console.log(canSum(7, [2,3]));
// console.log(canSum(7, [5,3,4,7]));
// console.log(canSum(7, [2,4]));
// console.log(canSum(8, [2,3,5]));
// console.log(canSum(300, [7,14]));

memo = {}
const canSum = (targetSum, numbers) => {
    if (targetSum in memo) return memo[targetSum];
    if (targetSum === 0) return true;
    if (targetSum < 0) return false;

    for(let num of numbers){
        const remainder = targetSum - num;
        if (canSum(remainder, numbers) === true){
            memo[targetSum] = true;
            return true;
        }
    }

    memo[targetSum] = false;
    return false;
};

console.log(canSum(7, [2,3]));
console.log(memo);